### PR TITLE
feat(ivy): add query inheritance

### PR DESCRIPTION
**Pull request by benlesh**
Adds inheritance handling for the following:

- viewQuery
- contentQueries
- contentQueriesRefresh

### CODE REVIEW

File: packages/core/src/render3/features/inherit_definition_feature.ts [hunk 1]
```diff
@@ -68,6 +68,53 @@ export function InheritDefinitionFeature(
         }
       }
 
+      // Merge View Queries
+      if (isComponentDef(definition) && isComponentDef(superDef)) {
+        const prevViewQuery = definition.viewQuery;
+        const superViewQuery = superDef.viewQuery;
+        if (superViewQuery) {
+          if (prevViewQuery) {
+            // tslint:disable-next-line:no-any viewQuery is readonly, but we need to set it
```
Comment (mhevery) [inline packages/core/src/render3/features/inherit_definition_feature.ts]: I don't see why you need this tslint disable. There must be something weird about your setup. We should investigate.

Preferred way is `(definition as {viewQuery: ViewQuery}).viewQuery = ...`

Also we could just take the `readonly` away?

File: packages/core/src/render3/features/inherit_definition_feature.ts [hunk 2]
```diff
@@ -68,6 +68,53 @@ export function InheritDefinitionFeature(
...
+            (definition as any).viewQuery = <T>(rf: RenderFlags, ctx: T): void => {
+              superViewQuery(rf, ctx);
+              prevViewQuery(rf, ctx);
+            };
+          } else {
+            // tslint:disable-next-line:no-any viewQuery is readonly, but we need to set it
```
Comment (mhevery) [inline packages/core/src/render3/features/inherit_definition_feature.ts]: same here.

File: packages/core/test/render3/Inherit_definition_feature_spec.ts [hunk 3]
```diff
@@ -307,6 +307,111 @@ describe('InheritDefinitionFeature', () => {
     expect(log).toEqual([['super', 1, 2], ['sub', 1, 2]]);
   });
 
+  it('should compose viewQuery', () => {
+    const log: Array<[string, RenderFlags, any]> = [];
+
+    // tslint:disable-next-line:class-as-namespace
```
Comment (mhevery) [inline packages/core/test/render3/Inherit_definition_feature_spec.ts]: revert `tslint`

File: packages/core/test/render3/Inherit_definition_feature_spec.ts [hunk 4]
```diff
@@ -307,6 +307,111 @@ describe('InheritDefinitionFeature', () => {
...
+        template: () => {},
+        selectors: [['', 'superDir', '']],
+        viewQuery: <T>(rf: RenderFlags, ctx: T) => {
+          log.push(['super', rf, ctx]);
+        },
+        factory: () => new SuperComponent(),
+      });
+    }
+
+    // tslint:disable-next-line:class-as-namespace
```
Comment (mhevery) [inline packages/core/test/render3/Inherit_definition_feature_spec.ts]: revert

File: packages/core/test/render3/Inherit_definition_feature_spec.ts [hunk 5]
```diff
@@ -307,6 +307,111 @@ describe('InheritDefinitionFeature', () => {
...
+
+    subDef.viewQuery !(1, context);
+
+    expect(log).toEqual([['super', 1, context], ['sub', 1, context]]);
+  });
+
+  it('should compose contentQueries', () => {
+    const log: string[] = [];
+
+    // tslint:disable-next-line:class-as-namespace
```
Comment (mhevery) [inline packages/core/test/render3/Inherit_definition_feature_spec.ts]: revert

File: packages/core/test/render3/Inherit_definition_feature_spec.ts [hunk 6]
```diff
@@ -307,6 +307,111 @@ describe('InheritDefinitionFeature', () => {
...
+    class SuperDirective {
+      static ngDirectiveDef = defineDirective({
+        type: SuperDirective,
+        selectors: [['', 'superDir', '']],
+        contentQueries: () => { log.push('super'); },
+        factory: () => new SuperDirective(),
+      });
+    }
+
+    // tslint:disable-next-line:class-as-namespace
```
Comment (mhevery) [inline packages/core/test/render3/Inherit_definition_feature_spec.ts]: revert

**Review by mhevery — APPROVED** [review]
Few nit picks but LGTM otherwise

File: packages/core/src/render3/features/inherit_definition_feature.ts (hunk 1 above)
Comment (benlesh) [reply packages/core/src/render3/features/inherit_definition_feature.ts]: > Also we could just take the readonly away?

We can, I just didn't want to do that if it was supposed to be readonly as a public API.

Comment (benlesh) [reply packages/core/src/render3/features/inherit_definition_feature.ts]: Done

File: packages/core/src/render3/features/inherit_definition_feature.ts (hunk 2 above)
Comment (benlesh) [reply packages/core/src/render3/features/inherit_definition_feature.ts]: Done

File: packages/core/test/render3/Inherit_definition_feature_spec.ts (hunk 5 above)
Comment (benlesh) [reply packages/core/test/render3/Inherit_definition_feature_spec.ts]: done

File: packages/core/test/render3/Inherit_definition_feature_spec.ts (hunk 6 above)
Comment (benlesh) [reply packages/core/test/render3/Inherit_definition_feature_spec.ts]: done

**Comment by benlesh** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:209452482:BASE:209452540:1534789322773:a53c125d)

**Comment by benlesh** [comment]
There's one Travis job that seems to be flaky.

**Pull request closed without merging**